espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/drivers/ComposeDriver.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.drivers
+
+import androidx.compose.ui.semantics.SemanticsProperties
+import androidx.compose.ui.test.*
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidElementStateException
+import io.appium.espressoserver.lib.handlers.exceptions.NoSuchElementException
+import io.appium.espressoserver.lib.helpers.findAllComposeNodeInteractions
+import io.appium.espressoserver.lib.helpers.getNodeInteractionById
+import io.appium.espressoserver.lib.model.*
+
+class ComposeDriver : AppDriver {
+    override val name = DriverContext.StrategyType.COMPOSE
+
+    override fun findElement(params: Locator): BaseElement {
+        val parentNodeInteraction =
+            params.elementId?.let { getNodeInteractionById(it) }
+        val nodeInteractions =
+            findAllComposeNodeInteractions(parentNodeInteraction, params)
+
+        if (nodeInteractions.fetchSemanticsNodes().isEmpty()) {
+            throw NoSuchElementException(
+                String.format(
+                    "Could not find compose element with strategy %s and selector %s",
+                    params.using, params.value
+                )
+            )
+        }
+        return ComposeElement(nodeInteractions[0])
+    }
+
+
+    override fun click(params: AppiumParams): Void? {
+        val nodeInteraction = getNodeInteractionById(params.elementId)
+        nodeInteraction.performClick()
+        return null
+    }
+
+    override fun getText(params: AppiumParams): String {
+        val nodeInteraction = getNodeInteractionById(params.elementId)
+        try {
+            return nodeInteraction.fetchSemanticsNode().config[SemanticsProperties.Text][0].text
+        } catch (e: IllegalStateException) {
```

Comment:
I would rather explicitly check if `.config[SemanticsProperties.Text]` is empty. Exception-controlled flows are not very stable, less performant and are prone to failures

Reply:
We will need two checks here in that case. first `.config[SemanticsProperties.Text]` call doesn't raise an exception and in case it doesn't, `.config[SemanticsProperties.Text]` should not be empty.

Reply:
not necessarily. You could use https://developer.android.com/reference/kotlin/androidx/compose/ui/semantics/SemanticsConfiguration?hl=fi#contains(androidx.compose.ui.semantics.SemanticsPropertyKey) to verify if the key is there (if not then throw illegal element state)
And then verify the length of the returned array. 
Is there any documentation on this property btw? I'm not quite sure why an array is returned as key value

Reply:
Looks like there is something related to merging the semantics properties. https://developer.android.com/jetpack/compose/semantics#merged-vs-unmerged

Also, from the code docs:
```
Absorb the semantic information from a child SemanticsNode into this configuration.
This merges the child's semantic configuration using the merge() method defined on the key. This is used when mergeDescendants is specified (for accessibility focusable nodes).
```

Reply:
this makes sense now. So, basically, we need to check key presence plus array length and throw illegal element state if any of these checks fails. Also, it would make sense to add the link above as a comment there, so one could better understand the logic flow there

Reply:
changed